README.md
If you're looking for a safe OpenGL abstraction, check out
[glium](https://github.com/glium/glium). This project is for people who want
total control over what happens and easily see what OpenGL functions are being
called.

## Goals

1. Implement OpenGL API with more type safety.
2. Everything *must* be zero-cost.
3. Rename OpenGL identifiers to comply with Rust's naming conventions.

Safety is not on the list because I don't trust myself enough to make that
judgement. It may be possible to make the API safe without violating the
zero-cost constraint.

It is literally **impossible** to prevent all OpenGL errors from occurring (like
glium promises) under the zero-cost constraint. If you want to have any
flexibility in your code (and you do) you must do what glium does: store the
owner of objects and synchronize access. I myself thought "oh nice I can use
lifetimes and borrowing concepts to ensure this and that" but objects with
lifetimes don't compose well and we can't create self referential structs
without unsafe code. Check out the design section if you're interested in the
details.

## Design

Designing a good API is hard. You have to balance ergonomics, flexibility and
performance. You're trying to guess what would be useful to potential consumers
of your library and abstract over potentially multiple, possibly yet unwritten
implementations.

### Tracking state

We don't try to prove that OpenGL is in the right state. Doing so for any
non-trivial program will incur some overhead, violating the zero-cost
constraint. Instead we make it harder to call OpenGL functions with the wrong
types.

The distinction between state and types is not very clear, since state can be
captured in the type. For example we could expose a type `UncompiledShader`

I'm having trouble writing the `compile_shader` definition, as it takes in a
`Shader<K, S>`, for which we do not care about the kind K nor the current
compile status S. On the OpenGL side the status may have changed and we need to
query it in order to know what it is. We could return a `Shader<K, Unknown>` and
add a function that takes that shader, does a glGetShaderiv call to query the
compile status and return a `Shader<K, CompileStatus>`. This keeps everything
type safe and zero-cost but it comes at a price: it adds a function to the API
that doesn't exist in OpenGL.

Another strategy is to simply query the status in `compile_shader`. Doing so
violates the zero-cost constraint because it does more than a regular
`glCompileShader` call. However, when would you ever want to

### Object relations

Do you see the problem with this example?

```rust
// Assume create_context creates an OpenGL context.
let gl1 = create_context();
let gl2 = create_context();
let n: ShaderName = gl1.create_shader(VERTEX_SHADER);
let mut s = Default::default();
gl2.get_shaderiv(&n, COMPILE_STATUS, &mut s);
```

We cannot prove an object name is valid without somehow storing the context it
was created from. This can be done at compile-time by creating a different type
for every context. 

```rust
let gl1 = create_context!(Gl1);
let gl2 = create_context!(Gl2);
let n: ShaderName<Gl1> = gl1.create_shader(VERTEX_SHADER);
let mut s = Default::default();
gl2.get_shaderiv(&n, COMPILE_STATUS, &mut s); // Compile error: Expected &ShaderName<Gl2>, got &ShaderName<Gl1>.
```

This doesn't work when we don't have an upper limit on the number of contexts we
might need. To ensure we dont' use an object name from a different context, we
must store something that uniquely identifies the context (like a reference),
and use it directly or check for equality with the used context.

```rust
struct ShaderName {
    gl: Rc<Gl>,
    name: NonZeroU32,
}
```

Obviously, this comes with a cost. This doesn't really matter much since it is
likely you'll only store a few shader names, if any. However, the same
relationship holds for all object names, like buffers, programs. The relation
also exists between programs and uniform locations.

It seems pretty silly right? What can we do? Well, we can try to limit the
possibility of making such errors by keeping related objects close together.

```rust
struct Program {
    name: ProgramName,
    // ...
    time_loc: UniformLocation<f32>,
    ambient_loc: UniformLocation<f32>,
}
```

With this setup it is still possible to do stupid things.

```rust
// Assume p1 and p2 are programs.
gl.uniform_1f(&p1, &p2.some_loc_not_from_p1, 0.5f);
```

In the end, if you write stupid code, it is going to do stupid things. We just
want to make it hard to do stupid things. Making it impossible to do stupid
things comes at a cost, which can be argued is small, but these small things add
up in a large application. I would rather have users of a library decide how
much protection they want. They should make the trade-off between flexibility,
correctness, speed, maintainability, etc.

### Abstracting over static and dynamic arrays

Should use the `array-generics` and `arrayvec` crates but it is hard to use them
well if you don't know what you are doing. Therefore I wrote [traits and
implementations](https://github.com/mickvangelderen/gl-typed-rust/blob/1e440ec4f643abca5e5ecd14d8af1665871a04f6/src/array.rs)
to abstract over statically and dynamically sized arrays. It works well but
might not be the best solution. As a bonus I took a look at the assembly to see
if it is worth doing at all.

```rust
impl GlTyped {
    // ...

    pub unsafe fn asm_shader_source<'a>(&self, name: &mut ShaderName, sources: &[&'a [u8]; 4]) {
        self.shader_source(name, sources);
    }

    pub unsafe fn asm_shader_source_slice<'a>(&self, name: &mut ShaderName, sources: &[&'a [u8]]) {
        self.shader_source(name, sources);
    }
}
```

Then I installed some utilities to make it easy to view the asm and count the lines of code.

```sh
cargo install cargo-asm
cargo install tokei
```

```sh
cargo asm gl_typed::GlTyped::asm_shader_source > array.s
cargo asm gl_typed::GlTyped::asm_shader_source_slice > slice.s
tokei -f *.s
--------------------------------------------------------------------------------
 Language             Files        Lines         Code     Comments       Blanks
--------------------------------------------------------------------------------
 Assembly                 2          278          278            0            0
--------------------------------------------------------------------------------
 array.s                              30           30            0            0
 slice.s                             248          248            0            0
--------------------------------------------------------------------------------
 Total                    2          278          278            0            0
--------------------------------------------------------------------------------
```

Here is the assembly for the `[&[u8]; 4]` version:

```asm
gl_typed::GlTyped::asm_shader_source:
 push    rbp
 mov     rbp, rsp
 and     rsp, -32
 sub     rsp, 96
 mov     r8, rdi
 mov     edi, dword, ptr, [rsi]
 movsd   xmm0, qword, ptr, [rdx, +, 16]
 movsd   xmm1, qword, ptr, [rdx]
 movlhps xmm1, xmm0
 movsd   xmm0, qword, ptr, [rdx, +, 48]
 movsd   xmm2, qword, ptr, [rdx, +, 32]
 movlhps xmm2, xmm0
 movaps  xmmword, ptr, [rsp, +, 48], xmm2
 movaps  xmmword, ptr, [rsp, +, 32], xmm1
 mov     ecx, dword, ptr, [rdx, +, 8]
 mov     esi, dword, ptr, [rdx, +, 24]
 mov     eax, dword, ptr, [rdx, +, 40]
 mov     edx, dword, ptr, [rdx, +, 56]
 mov     dword, ptr, [rsp, +, 16], ecx
 mov     dword, ptr, [rsp, +, 20], esi
 mov     dword, ptr, [rsp, +, 24], eax
 mov     dword, ptr, [rsp, +, 28], edx
 lea     rdx, [rsp, +, 32]
 lea     rcx, [rsp, +, 16]
 mov     esi, 4
 call    qword, ptr, [r8, +, 7712]
 mov     rsp, rbp
 pop     rbp
 ret
```

Even with all the code to deal with types that might drop, functions that might
panic, and assertions all over the place the code can compile down to a
relatively small number of instructions. I'm not proficient in asm but it looks
like the byte slices are effectively split pointers and lengths. The loop is
unrolled, there seems to be some simd instructions and there are no branches at
all.

### Do I need a transmute trait?

`S: Transmute<T>` guarantees that `S` and `T` are of the same size and have the
same valid bit patterns.

Since the reverse must also be true, `T: Transmute<S>` should be implemented
when `S: Transmute<T>`.

We can have TransmuteFrom and TransmuteInto traits to allow one-directional
transmutes (for sub types). `Transmute<T>` should then imply `TransmuteFrom<T>`
and `TransmuteInto<T>`. When `S: TransmuteFrom<T>`, all `T` as valid `S`. Thus
we can cast `&T` to `&S` but not `&mut T` to `&mut S`! If we cast `&mut T` to
`&mut S` we can write any S to it, which may not be a valid T. We can however,
with the imaginary `&wo` write only mutable reference, cast `&wo S` to `&wo T`
since writing a `T` to an `S` is fine.

| every `S` is a valid `T` | every `T` is a valid `S` | `&S as &T` | `&mut S as &mut T` | `&wo S as &wo T` |
| --- | --- | --- | --- | --- |
| no | yes | no | no | yes |
| yes | no | yes | no | no |
| yes | yes | yes | yes | yes |

Probably for good reason but unfortunately, `AsMut<T>` is not implemented for
all `T`. For example `i32` does not implement `AsMut<i32>`. That means that we
can't have `Transmute<T>: AsMut<T>` and use the standard trait since we can't
implement `AsMut` for `i32` ourselves. We would have to provide a wrapper type.

For now the `From`, `Into`, `AsRef` and `AsMut` functions are prefixed with
`transmute_`. Even though it's unfortunate we have to provide implementations
for the standard traits and then again for transmute, it works.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
[Unreleased]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.5.0...HEAD

## [0.6.0] - 2019-05-27
[0.6.0]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.5.0...0.6.0

### Added
 - A bunch of direct state access functions.
 
### Changed
 - Downgrade OpenGL from 4.6 to 4.5.
 - Panic if the driver returns an invalid enum value.
 - Unwrap by default, provide try_ variants of functions to explicitly handle errors.
 - No longer provide the ability to write into a memory location so we can perform checks on values returned by OpenGL.
 
### Removed
 - Convute dependency.
 - Custom symbols, only the OpenGL constants are created now.
 - Multi-object creation and deletion has been removed because they shouldn't
   happen in performance critical situations. It is hard to allocate a bunch of
   them anyway without doing your own sub-allocation. In that case use the raw
   OpenGL api.
 - Deprecated a whole bunch of non direct state access functions.

## [0.5.0] - 2019-05-08
[0.5.0]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.8...0.5.0

### Removed
 - **breaking** `TextureTargetGE2D` and `TextureTargetGE3D`. The benefits are not worth the added complexity. 

## [0.4.8] - 2019-05-08
[0.4.8]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.7...0.4.8

### Added

 - `glClipControl`
 - `GL_DEPTH_ATTACHMENT` and `GL_STENCIL_ATTACHMENT` to `FramebufferAttachment`
 
## [0.4.7] - 2019-05-03
[0.4.7]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.6...0.4.7

### Added
 - `get_programiv_move`
 - `get_program_info_log_move`

## [0.4.6] - 2019-04-26
[0.4.6]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.5...0.4.6

### Added
 - Extention ARB_texture_filter_anisotropic.
 - Sampler objects.

## [0.4.5] - 2019-04-25
[0.4.5]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.4...0.4.5

### Changed
 - Fixed build issues, version 0.4.4 should never be used.

## [0.4.4] - 2019-04-25 (broken)
[0.4.4]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.3...0.4.4

### Added
 - glClearDepth
 - Renderbuffer functionality
 - glColorMask
 - glDepthMask
 - glDepthFunc
 - glDepthRange
 - glStencilMask
 
### Changed
 - Removed unused type parameter from `delete_buffers`.
 - Made locations' new_unchecked const functions.

## [0.4.3] - 2019-04-10
[0.4.3]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.2...0.4.3

### Added

 - Uniform buffer object functionality.

## [0.4.2] - 2019-04-10
[0.4.2]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.1...0.4.2

### Added

 - `unuse_program`
 - `get_context_flags`
 - `draw_buffers`

## [0.4.1] - 2019-04-05
[0.4.1]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.4.0...0.4.1

### Added
 - `get_error`
 - Multiple uniform setters `uniform_{1|2|3|4}{i|ui|f}`.

## [0.4.0] - 2019-04-03
[0.4.0]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.3.0...0.4.0

### Added
 - `is_some` and `is_none` to `Option{Attribute|Uniform}Location`.
 - `From<Option{X}Location> for `Option<{X}Location>` where `X = Attribute|Uniform`.
 - 'OptionAttributeLocation::NONE' and `OptionUniformLocation::NONE`.

### Changed
 - `Gl::get_attrib_location` now returns `OptionAttributeLocation` instead of
   `Option<AttributeLocation>`.
 - Copied `std::convert::Try{From|Into}` into this crate until they're stable. I
   thought they were but it turns out I cant actually use them with the current
   stable rustc.

### Removed
 - Warnings about aliases being ignored in build.

## [0.3.0] - 2019-03-12
[0.3.0]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.2.0...0.3.0

The conversion traits were moved into a stand-alone crate `convute`.

### Changed
 - The wrap and unwrap traits have been replaced with the more solid traits from
   `convute`. The traits are re-exported in the `convert` and `marker` modules.

## [0.2.0] - 2019-03-08
[0.2.0]: https://github.com/mickvangelderen/gl-typed-rust/compare/v0.1.2...0.2.0

This breaking update is pretty invasive because almost all calls now take
parameters by value instead of reference. The best part in this release are the
conversion traits `TryUnwrapAll` and friends.

Tracking ownership of Name objects is better left to a higher abstraction layer.
We can therefore make the types Copy and provide some zero-copy array
conversions.

We provide custom types for `Option<AttributeLocation>` and
`Option<UniformLocation>` so we can store them with zero memory overhead.
Unfortunately that means we have to convert them into options when we use them
which is not very ergonomic. The previous solutions was modify the values so
that the sentinel values become zero, and storing them in a NonZero* type.
Unfortunately that does not allow us to transmute between the super and the
subtype and thus prevent the zero-copy array conversions.

Usually you don't want to bind or unbind dynamically, you know at compile-time
which one you want so then it makes sense for them to be distinct methods.

### Added
 - Conversions between owned arrays and slices of `Option<T>` and `T` where they
   are of the same size.
 - Specialized types for locations wrapped in options: `OptionAttributeLocation`
   and `OptionUniformLocation`. We no longer utilize NonZero* to get the memory
   optimization. You can convert them into an actual option using `into_option`
   where they will actually be used.
 - Specialized unbind functions where it makes sense.

### Changed
 - Many types are now `Copy` and provide `into_*` function instead of `as_*`.
 - Unbinding is no longer done through the bind functions but through a
   specialized function.

### Removed

src/array.rs
pub trait Array {
    type Item;

    fn is_empty(&self) -> bool;

    fn len(&self) -> usize;

    fn as_slice(&self) -> &[Self::Item];

    fn as_mut_slice(&mut self) -> &mut [Self::Item];

    fn as_ptr(&self) -> *const Self::Item;

    fn as_mut_ptr(&mut self) -> *mut Self::Item;
}

pub trait ArrayMap<ItemOut>: Array {
    type ArrayOut: Array<Item = ItemOut>;

    fn map<F: FnMut(&Self::Item) -> ItemOut>(&self, f: F) -> Self::ArrayOut;
}

impl<T> Array for [T] {
    type Item = T;

    #[inline]
    fn is_empty(&self) -> bool {
        self.len() == 0
    }

    #[inline]
    fn len(&self) -> usize {
        <[T]>::len(self)
    }

    #[inline]
    fn as_slice(&self) -> &[T] {
        &*self
    }

    #[inline]
    fn as_mut_slice(&mut self) -> &mut [T] {
        &mut *self
    }

    #[inline]
    fn as_ptr(&self) -> *const T {
        <[T]>::as_ptr(self)
    }

    #[inline]
    fn as_mut_ptr(&mut self) -> *mut T {
        <[T]>::as_mut_ptr(self)
    }
}

impl<T, U> ArrayMap<U> for [T] {
    type ArrayOut = Vec<U>;

    #[inline]
    fn map<F: FnMut(&Self::Item) -> U>(&self, f: F) -> Self::ArrayOut {
        self.iter().map(f).collect()
    }
}

impl<T> Array for Vec<T> {
    type Item = T;

    #[inline]
    fn is_empty(&self) -> bool {
        self.len() == 0
    }

    #[inline]
    fn len(&self) -> usize {
        Vec::len(self)
    }

    #[inline]
    fn as_slice(&self) -> &[T] {
        Vec::as_slice(self)
    }

    #[inline]
    fn as_mut_slice(&mut self) -> &mut [T] {
        Vec::as_mut_slice(self)
    }

    #[inline]
    fn as_ptr(&self) -> *const T {
        <[T]>::as_ptr(self)
    }

    #[inline]
    fn as_mut_ptr(&mut self) -> *mut T {
        <[T]>::as_mut_ptr(self)
    }
}

impl<T, U> ArrayMap<U> for Vec<T> {
    type ArrayOut = Vec<U>;

    #[inline]
    fn map<F: FnMut(&Self::Item) -> U>(&self, f: F) -> Self::ArrayOut {
        self.iter().map(f).collect()
    }
}

macro_rules! impl_array {
    ($($N:expr,)+) => {
        $(
            impl<T> Array for [T; $N] {
                type Item = T;

                #[inline]
                fn is_empty(&self) -> bool {
                    self.len() == 0
                }

                #[inline]
                fn len(&self) -> usize {
                    <[T]>::len(self)
                }

                #[inline]
                fn as_slice(&self) -> &[T] {
                    self
                }

                #[inline]
                fn as_mut_slice(&mut self) -> &mut [T] {
                    self
                }

                #[inline]
                fn as_ptr(&self) -> *const T {
                    <[T]>::as_ptr(self)
                }

                #[inline]
                fn as_mut_ptr(&mut self) -> *mut T {
                    <[T]>::as_mut_ptr(self)
                }
            }

            impl<T, U> ArrayMap<U> for [T; $N] {
                type ArrayOut = [U; $N];

                #[inline]
                fn map<F: FnMut(&Self::Item) -> U>(&self, mut f: F) -> Self::ArrayOut {
                    use std::mem;
                    use std::ptr;

                    unsafe {
                        struct Initializer<A: Array> {
                            index: usize,
                            array: mem::ManuallyDrop<A>,
                        }

                        impl<A: Array> Initializer<A> {
                            #[inline]
                            pub fn push(&mut self, item: A::Item) {
                                assert!(self.index < self.array.len());
                                (self.array.as_mut_slice())[self.index] = item;
                                self.index += 1;
                            }

                            #[inline]
                            pub fn finish(self) -> A {
                                assert_eq!(self.index, self.array.len());
                                unsafe {
                                    let mut initializer = mem::ManuallyDrop::new(self);
                                    let array = mem::replace(&mut initializer.array, mem::uninitialized());
                                    mem::ManuallyDrop::into_inner(array)
                                }
                            }
                        }

                        impl<A: Array> Drop for Initializer<A> {
                            #[inline]
                            fn drop(&mut self) {
                                // Drop all values initialized so far.
                                unsafe {
                                    let s = self.array.as_mut_slice();
                                    for i in 0..self.index {
                                        ptr::drop_in_place(&mut s[i])
                                    }
                                }
                            }
                        }

                        let mut init = Initializer::<Self::ArrayOut> {
                            index: 0,
                            array: mem::uninitialized(),
                        };

                        for item in self {
                            init.push(f(item))
                        }

                        Initializer::finish(init)
                    }
                }
            }
        )+
    };
}

impl_array! {
    0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
    10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
    20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
    30, 31, 32,
}
